app/controllers/solidus_paypal_commerce_platform/shipping_rates_controller.rb

```diff
@@ -7,18 +7,38 @@ class ShippingRatesController < ::Spree::Api::BaseController
 
     def simulate_shipping_rates
       authorize! :show, @order, order_token
-      order_simulator = SolidusPaypalCommercePlatform.config.order_simulator_class.new(@order)
-      simulated_order = order_simulator.simulate_with_address(params[:address])
 
-      if simulated_order.ship_address.valid?
-        render json: SolidusPaypalCommercePlatform::PaypalOrder.new(simulated_order).to_replace_json, status: :ok
-      else
-        render json: simulated_order.ship_address.errors.full_messages, status: :unprocessable_entity
+      @order.transaction do
+        build_address(params[:address])
+        @order.ensure_updated_shipments
+        @order.email = "info@solidus.io" unless @order.email
+        @order.contents.advance
+        @paypal_order = SolidusPaypalCommercePlatform::PaypalOrder.new(@order).to_replace_json
+        raise ActiveRecord::Rollback
       end
+
+      render json: @paypal_order, status: :ok
     end
 
     private
 
+    def build_address(address)
+      country = ::Spree::Country.find_by(iso: address[:country_code])
+      # Also adds fake information for a few fields, so validations can run
+      new_ship_address = ::Spree::Address.new(
+        city: address[:city],
+        state: country.states.find_by(abbr: address[:state]),
+        state_name: address[:state],
+        zipcode: address[:postal_code],
+        country: country,
+        address1: "123 Fake St.",
+        phone: "123456789",
+        firstname: "Fake"
+      )
+
+      @order.update(ship_address: new_ship_address)
+    end
+
```

Comment:
What about moving most of this to an external object and keep the controller a little slender?

Ideally the action should look more or less like this:

```rb
@paypal_order = SolidusPaypalCommercePlatform::PaypalOrder.new(@order)

render json: @paypal_order.to_replace_json(address: params[:address])
```

Reply:
I moved the address update functionality to SPCP::PayPalAddress, since its responsibility is to format paypal addresses to Spree addresses. How does that look?